### ISSUE TITLE
Is no IV allowed

### ISSUE THREAD

**Issue by tylerlong** [krzyzanowskim/CryptoSwift#360]
Hi,

I am trying to translate the following C# code into Swift 3:

```csharp
        private object Decrypt(string dataString)
        {
            var key = Convert.FromBase64String(subscriptionInfo.deliveryMode.encryptionKey);
            var keyParameter = ParameterUtilities.CreateKeyParameter("AES", key);
            var cipher = CipherUtilities.GetCipher("AES/ECB/PKCS7Padding");
            cipher.Init(false, keyParameter);

            var data = Convert.FromBase64String(dataString);
            var memoryStream = new MemoryStream(data, false);
            var cipherStream = new CipherStream(memoryStream, cipher, null);

            var bufferSize = 1024;
            var buffer = new byte[bufferSize];
            var length = 0;
            var resultStream = new MemoryStream();
            while ((length = cipherStream.Read(buffer, 0, bufferSize)) > 0)
            {
                resultStream.Write(buffer, 0, length);
            }
            var resultBytes = resultStream.ToArray();
            var result = Encoding.UTF8.GetString(resultBytes, 0, resultBytes.Length);
            return JsonConvert.DeserializeObject(result);
        }
```

You can see that, it is AES, ECB, PKCS7Padding.

I tried the following swift 3 code

```swift
        let base64Message = message.data.message as! String
        let encrypted = Array(base64Message.utf8)
        let padded = PKCS7().add(to: encrypted, blockSize: AES.blockSize)
        let key: [UInt8] = Array(subscriptionInfo!.deliveryMode!.encryptionKey!.utf8)
        let iv: [UInt8] = AES.randomIV(AES.blockSize)
        let decrypted = try! AES(key: key, iv: iv, blockMode: .ECB, padding: PKCS7()).decrypt(padded)
        let result = String(data: Data(bytes: decrypted), encoding: String.Encoding.utf8)!
        for listener in listeners {
            listener(result)
        }
```

I think it is because of `IV`. In C# code, there is no `IV` at all. But in swift, if I don't provide one, a default one will be used: https://github.com/krzyzanowskim/CryptoSwift/blob/master/Sources/CryptoSwift/AES.swift#L119

A default IV is different from no IV. I am wondering how could I make the swift code same as the C# one.

By the way, I have to do ` let padded = PKCS7().add(to: encrypted, blockSize: AES.blockSize)` in swift which is not required in C#. That might be the reason it doesn't work too.

**Comment by krzyzanowskim** [comment]
ECB don't use IV so It won't be used even though the default IV is set there. You can use just `nil` as value for ECB, the parameter is optional.

You shouldn't do manual padding here if you use PKCS7() as AES a parameter. Why did you have add it manually? 

btw. do you get right result? It looks to me that you handle Base64 in the wrong way in this snippet.

**Comment by tylerlong** [comment]
@krzyzanowskim Thank you so much for your response.

> You shouldn't do manual padding here if you use PKCS7() as AES a parameter. Why did you have add it manually?

Without the following line:

```swift
let padded = PKCS7().add(to: encrypted, blockSize: AES.blockSize)
```

The following error is thrown:

```
fatal error: 'try!' expression unexpectedly raised an error: CryptoSwift.AES.Error.dataPaddingRequired: file /Library/Caches/com.apple.xbs/Sources/swiftlang/swiftlang-800.0.46.2/src/swift/stdlib/public/core/ErrorType.swift, line 178
```

If I do padding manually, I can get result as `[UInt8]`, but the data is incorrect because I cannot convert it to string (invalid UTF-8 sequence I think).

**Comment by tylerlong** [comment]
OOPS! You are right, Base64 to [UInt8] shouldn't be handled like that. 

Let me try again and update you later!

**Comment by tylerlong** [comment]
I have got it working:

```swift
        let base64Message = message.data.message as! String
        let encrypted = Data(base64Encoded: base64Message)!.bytes
        let key = Data(base64Encoded: subscriptionInfo!.deliveryMode!.encryptionKey!)!.bytes
        let decrypted = try! AES(key: key, iv: nil, blockMode: .ECB, padding: PKCS7()).decrypt(encrypted)
        let result = String(bytes: decrypted, encoding: String.Encoding.utf8)!
```

Base64 string is not normal string, I shouldn't convert it to bytes using UTF-8 encoding.
